src/bidiMapper/domains/context/logEntry.ts

```diff
@@ -0,0 +1,276 @@
+    if (
+      args[0].type === 'string' &&
+      this._isFormmatSpecifier(args[0].value.toString()) &&
+      formatText
+    ) {
+      return this._formatter(args);
+    }
+
+    // if args[0] is not a format specifier, just join the args with \u0020
+    return args
+      .map((arg) => (arg as { type: string; value: any }).value.toString())
+      .join('\u0020');
+  }
+
+  // convert CDP StackTrace object to Bidi StackTrace object
+  private _getBidiStackTrace(
+    cdpStackTrace: Protocol.Runtime.StackTrace | undefined
+  ): Script.StackTrace | undefined {
+    const stackFrames = cdpStackTrace?.callFrames.map((callFrame) => {
+      return {
+        columnNumber: callFrame.columnNumber,
+        functionName: callFrame.functionName,
+        lineNumber: callFrame.lineNumber,
+        url: callFrame.url,
+      };
+    });
+
+    const bidiStackTrace = stackFrames
+      ? {
+          callFrames: stackFrames,
+        }
+      : undefined;
+
+    return bidiStackTrace;
+  }
+
+  private _initializeLogEntryAddedEventListener() {
+    this._cdpClient.Runtime.on('consoleAPICalled', async (params) => {
+      const args = await Promise.all(
+        params.args.map(async (arg) => {
+          return await this._serializeCdpObject(arg);
+        })
+      );
+
+      await this._bidiServer.sendMessage({
+        method: 'log.entryAdded',
+        params: {
+          level: this._getLogLevel(params.type),
+          text: this._getRemoteValuesText(args, true),
+          timestamp: params.timestamp,
+          stackTrace: this._getBidiStackTrace(params.stackTrace),
+          type: 'console',
+          method: params.type,
+          realm: this._contextId,
+          args: args,
+        },
+      });
+    });
+  }
+
+  // TODO sadym: `dummyContextObject` needed for the running context.
+  // Use the proper `executionContextId` instead:
+  // https://github.com/GoogleChromeLabs/chromium-bidi/issues/52
+  private async _getDummyContextId(): Promise<string> {
+    const dummyContextObject = await this._cdpClient.Runtime.evaluate({
+      expression: '(()=>{return {}})()',
+    });
+    return dummyContextObject.result.objectId!;
+  }
+
+  /**
+   * Serializes a given CDP object into BiDi, keeping references in the
+   * target's `globalThis`.
+   * @param cdpObject CDP remote object to be serialized.
+   */
+  private async _serializeCdpObject(
+    cdpObject: Protocol.Runtime.RemoteObject
+  ): Promise<CommonDataTypes.RemoteValue> {
+    const response = await this._cdpClient.Runtime.callFunctionOn({
+      functionDeclaration: `${this.EVALUATOR_SCRIPT}.serialize`,
+      objectId: await this._getDummyContextId(),
+      arguments: [cdpObject],
+      returnByValue: true,
+    });
+
+    if (response.exceptionDetails)
+      // Serialization failed unexpectidely.
+      throw new Error(
+        'Cannot serialize object: ' + response.exceptionDetails.text
+      );
+
+    return response.result.value;
+  }
+}
```

Comment:
I cannot say I like the copy-paste of such a tricky logic which will be replaced with another one (after we move serialization to CDP).

Reply:
removed the duplicate codes.